plugins/fileservices/fileservices.cpp

```diff
@@ -2509,4 +2511,66 @@ FILESERVICES_API void FILESERVICES_CALL fsDfuPlusExec(ICodeContext * ctx,const c
     }
 }
 
+FILESERVICES_API const char * FILESERVICES_CALL fsGetEspURL(const char *username, const char *userPW)
+{
+    Owned<IConstEnvironment> daliEnv = openDaliEnvironment();
+    Owned<IPropertyTree> env = getEnvironmentTree(daliEnv);
+
+    if (env.get())
+    {
+        StringBuffer espURL;
+        StringBuffer espInstanceComputerName;
+        StringBuffer bindingProtocol;
+        StringBuffer xpath;
+        StringBuffer instanceAddress;
+        StringBuffer espServiceType;
+        StringBuffer credentials;
+
+        Owned<IPropertyTreeIterator> espProcessIter = env->getElements("Software/EspProcess");
+        ForEach(*espProcessIter)
+        {
+            Owned<IPropertyTreeIterator> espBindingIter = espProcessIter->query().getElements("EspBinding");
+            ForEach(*espBindingIter)
+            {
+                espBindingIter->query().getProp("@service",espURL.clear());
+                xpath.setf("Software/EspService[@name=\"%s\"]/Properties/@type", espURL.str());
+
+                if(env->getProp(xpath.str(), espServiceType.clear()))
+                {
+                    if (strieq(espServiceType.str(),"WsSMC"))
+                    {
+                        if (espBindingIter->query().getProp("@protocol",bindingProtocol.clear()))
+                        {
+                            Owned<IPropertyTreeIterator> espInstanceIter = espProcessIter->query().getElements("Instance");
+                            ForEach(*espInstanceIter)
+                            {
+                                if (espInstanceIter->query().getProp("@computer",espInstanceComputerName.clear()))
+                                {
+                                    xpath.setf("Hardware/Computer[@name=\"%s\"]/@netAddress",espInstanceComputerName.str());
+                                    if (env->getProp(xpath.str(),instanceAddress.clear()))
```

Comment:
I think a lot of this code should really be encapsulated inside the environment.hpp interfaces.
Some of it may already be possible (e.g. getInstance), but should be extended where not, so you can use those interfaces to get down to the part of the environment that's interesting.

Reply:
This new function was heavily cribbed from getAccessibleEspServerURL() in that same file. If most of the functionality is moved to environment.hpp then we should probably also spawn a Jira to refactor getAccessibleEspServerURL() afterwards.

Reply:
Might make sense to merge this as is and have a Jira for refactoring both this and getAccessibleEspServerURL

Reply:
I like that idea, as it gives the refactoring work two working targets.